### PR TITLE
Dynmap compatability

### LINKED ISSUE #728: [Suggestion] Please add support for Dynmap
Please add support for Dynmap, something along the lines of the attached image. 
image I have found documentation and they do have an API although I do not know how useful their documentation is, here are some links that might help. https://gist.github.com/JOO200/eb99b294ebd16a353750f83d8dcdce23 (all the API documentation I could find. Note: was given to me on their discord server.) https://github.com/webbukkit/dynmap/wiki/Using-markers#area-markers (Might help with getting the area markers for the colony borders.)
Note: Dynmap is different from JourneyMap as it is targeted towards servers I believe. This is because it runs in a web browser without needing to have the Minecraft client open. It also allows Isometric map viewing.
Link to Dynmap Forge Version: https://www.curseforge.com/minecraft/mc-mods/dynmapforge

Thanks.

### PR DESCRIPTION
**Pull request by Thodor12**
Closes ldtteam/minecolonies-features#728

# Changes proposed in this pull request:
- Ability to render borders of colonies on Dynmap using a new marker set "Colonies".
- Renders the colony name, owner and population count on the description of the colony.
- Automatically updates Dynmap when the following occur:
  - The colony name gets changed
  - The colony it's team color is changed
  - The colony borders change when a building is constructured/repaired/removed
  - The colony it's population count changes, either when colonists are born or die.

What is not added in the pull request:
- Ability for Minecolonies specific blocks (primarly hut blocks) to render on Dynmap
- Ability for DO blocks to render on Dynmap (this would have to be implemented from within DO, should not appear here)

Review please

### CODE REVIEW

**Comment by IceGoalie31** [comment]
Thank you so much

File: src/api/java/com/minecolonies/api/compatibility/dynmap/ColonyChunkClaimCalculator.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import com.minecolonies.api.IMinecoloniesAPI;
+import com.minecolonies.api.colony.IColony;
+import com.minecolonies.api.colony.buildings.IBuilding;
+import net.minecraft.world.level.ChunkPos;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+public class ColonyChunkClaimCalculator
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/compatibility/dynmap/ColonyChunkClaimCalculator.java:17]: javadoc on classes and public methods

File: src/api/java/com/minecolonies/api/compatibility/dynmap/ColonyChunkClaimCalculator.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        // Claim all the buildings
+        colony.getBuildingManager().getBuildings().values().forEach(f -> claimBuilding(f, claimedChunks));
+
+        return claimedChunks;
+    }
+
+    private static void getProtectionArea(final IColony colony, Set<ChunkPos> claimedChunks)
+    {
+        var protectionAreaCenterChunk = new ChunkPos(colony.getCenter());
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/compatibility/dynmap/ColonyChunkClaimCalculator.java]: Please avoid var if possible

File: src/api/java/com/minecolonies/api/compatibility/dynmap/Constants.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.minecolonies.api.compatibility.dynmap;
+
+public final class Constants
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/compatibility/dynmap/Constants.java]: Id call the class DynmapConstants + javadoc

File: src/api/java/com/minecolonies/api/compatibility/dynmap/DynmapApiListener.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package com.minecolonies.api.compatibility.dynmap;
+
+import com.minecolonies.api.util.Log;
+import org.dynmap.DynmapCommonAPI;
+import org.dynmap.DynmapCommonAPIListener;
+
+import java.util.function.Consumer;
+
+public class DynmapApiListener extends DynmapCommonAPIListener
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/compatibility/dynmap/DynmapApiListener.java:15]: javadoc on public methods and classes

File: src/api/java/com/minecolonies/api/compatibility/dynmap/DynmapCompat.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.minecolonies.api.compatibility.dynmap;
+
+public class DynmapCompat extends DynmapProxy
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/compatibility/dynmap/DynmapCompat.java:6]: jdoc

File: src/api/java/com/minecolonies/api/compatibility/dynmap/DynmapIntegration.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+import com.minecolonies.api.compatibility.dynmap.area.AreaGenerator;
+import net.minecraft.world.level.storage.ServerLevelData;
+import org.dynmap.markers.MarkerAPI;
+import org.dynmap.markers.MarkerSet;
+
+import java.awt.*;
+
+import static com.minecolonies.api.compatibility.dynmap.Constants.*;
+
+public class DynmapIntegration
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/compatibility/dynmap/DynmapIntegration.java:24]: jdoc on class and public methods

File: src/api/java/com/minecolonies/api/compatibility/dynmap/DynmapIntegration.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+{
+    private final MarkerSet colonySet;
+
+    public DynmapIntegration(MarkerAPI markerApi)
+    {
+        var colonySetInternal = markerApi.getMarkerSet(DYNMAP_COLONY_MARKER_SET_ID);
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/compatibility/dynmap/DynmapIntegration.java]: Also please less var

File: src/api/java/com/minecolonies/api/compatibility/Compatibility.java [hunk 8]
```diff
@@ -296,4 +301,72 @@ public static List<ItemStack> getCombsFromHive(BlockPos pos, Level world, int am
...
+     *
+     * @param colony The colony
+     */
+    public static void updateColonyName(final IColony colony)
+    {
+        dynmapCompat.updateColonyName(colony);
+    }
+
+    /**
+     * Update the colony team color
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/compatibility/Compatibility.java]: Jdoc should probably be "Called on updating the colony team color"

File: src/main/java/com/minecolonies/coremod/network/messages/server/colony/ColonyDeleteOwnMessage.java [hunk 9]
```diff
@@ -50,6 +51,10 @@ public void onExecute(final NetworkEvent.Context ctxIn, final boolean isLogicalS
         {
             IColonyManager.getInstance().deleteColonyByDimension(colony.getID(), false, colony.getDimension());
             MessageUtils.format(MESSAGE_INFO_COLONY_DELETE_SUCCESS).sendTo(player);
+
+            if (isLogicalServer) {
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/network/messages/server/colony/ColonyDeleteOwnMessage.java]: brackets, also applies to the next

**Review by Raycoms — CHANGES_REQUESTED** [review]
Implementation wise is looks okay, style wise it needs some cleanup

**Comment by someaddons** [comment]
Did you try loading minecolonies without dynmap present yet?

**Comment by Thodor12** [comment]
> Did you try loading minecolonies without dynmap present yet?

Yes, this works just fine

**Review by someaddons — APPROVED** [review]

**Pull request merged**